File: ios/MoproApp/rapidsnark.swift
#if canImport(rapidsnarkC)
import rapidsnarkC
#endif

import Darwin.C.string
import Foundation

public let defaultProofBufferSize = 1024
public let defaultErrorBufferSize = 256

/**
 Performs cryptographic proofs using the Groth16 proving scheme
 
 - Parameters:
 - zkey: The zkey data used to generate the proof.
 - witness: The witness data used to generate the proof.
 - proofBufferSize: The size of the buffer to store the proof. Defaults to a predefined value.
 - publicBufferSize: The size for the public signal buffer. If not provided, it's calculated dynamically.
 - errorBufferSize: The size of the buffer for error messages. Defaults to a predefined value.
 
 - Throws: `RapidsnarkProverError` child classes
 if the proof generation fails, with the error message indicating the reason for failure.
 
 - Returns: A tuple containing the generated proof as a string and the public signals as a string.
 */
public func groth16Prove(
    zkey: Data,
    witness: Data,
    proofBufferSize: Int = defaultProofBufferSize,
    publicBufferSize: Int? = nil,
    errorBufferSize: Int = defaultErrorBufferSize
) throws -> (proof: String, publicSignals: String) {
    let zkeyBuf = NSData(data: zkey).bytes
    let witnessBuf = NSData(data: witness).bytes
    
    // calculate the size for the public signal buffer if not provided
    var currentPublicBufferSize : Int;
    if let publicBufferSize {
        currentPublicBufferSize = publicBufferSize;
    } else {
        currentPublicBufferSize = try groth16PublicSizeForZkeyBuf(zkey: zkey, errorBufferSize: errorBufferSize);
    }
    
    var proofBuffer = Array<CChar>(repeating: 0, count: proofBufferSize);
    var publicBuffer = Array<CChar>(repeating: 0, count: currentPublicBufferSize)
    var errorMessageBuffer: [CChar] = Array(repeating: 0, count: errorBufferSize)
    
    var proofBufferSizeUInt = UInt(proofBufferSize)
    var currentPublicBufferSizeUInt = UInt(currentPublicBufferSize)
    let errorBufferSizeUInt = UInt(errorBufferSize)
    
    // Call the rapidsnark C++ library function to perform the Groth16 proof
    let statusCode = groth16_prover(
        zkeyBuf, UInt(zkey.count),
        witnessBuf, UInt(witness.count),
        &proofBuffer, &proofBufferSizeUInt,
        &publicBuffer, &currentPublicBufferSizeUInt,
        &errorMessageBuffer, errorBufferSizeUInt
    );
    
    if (statusCode == PROVER_OK) {
        return (String(cString: proofBuffer), String(cString: publicBuffer))
    }
    
    throw groth16proverStatusCodeErrors(statusCode, message: String(cString: errorMessageBuffer))
}

/**
 Performs cryptographic proofs using the Groth16 proving scheme.
 
 - Parameters:
 - zkeyPath: The path to the .zkey file.
 - witness: The witness data used to generate the proof.
 - proofBufferSize: The size of the buffer to store the proof. Defaults to a predefined value.
 - publicBufferSize: The size for the public signal buffer. If not provided, it's calculated dynamically.
 - errorBufferSize: The size of the buffer for error messages. Defaults to a predefined value.
 
 - Throws: `RapidsnarkProverError` child classes
 if the proof generation fails, with the error message indicating the reason for failure.
 
 - Returns: A tuple containing the generated proof as a string and the public signals as a string.
 */
public func groth16ProveWithZKeyFilePath(
    zkeyPath: String,
    witness: Data,
    proofBufferSize: Int = defaultProofBufferSize,
    publicBufferSize: Int? = nil,
    errorBufferSize: Int = defaultErrorBufferSize
) throws -> (proof: String, publicSignals: String) {
    let witnessBuf = NSData(data: witness).bytes
    
    // calculate the size for the public signal buffer if not provided
    var currentPublicBufferSize : Int;
    if let publicBufferSize {
        currentPublicBufferSize = publicBufferSize;
    } else {
        currentPublicBufferSize = try groth16PublicSizeForZkeyFile(zkeyPath: zkeyPath, errorBufferSize: errorBufferSize);
    }
    
    var proofBuffer = Array<CChar>(repeating: 0, count: proofBufferSize);
    var publicBuffer = Array<CChar>(repeating: 0, count: currentPublicBufferSize)
    var errorMessageBuffer: [CChar] = Array(repeating: 0, count: errorBufferSize)
    
    var proofBufferSizeUInt = UInt(proofBufferSize)
    var currentPublicBufferSizeUInt = UInt(currentPublicBufferSize)
    let errorBufferSizeUInt = UInt(errorBufferSize)
    
    // Call the rapidsnark C++ library function to perform the Groth16 proof
    let statusCode = groth16_prover_zkey_file(
        zkeyPath,
        witnessBuf, UInt(witness.count),
        &proofBuffer, &proofBufferSizeUInt,
        &publicBuffer, &currentPublicBufferSizeUInt,
        &errorMessageBuffer, errorBufferSizeUInt
    );
    
    if (statusCode == PROVER_OK) {
        return (String(cString: proofBuffer), String(cString: publicBuffer))
    }
    
    throw groth16proverStatusCodeErrors(statusCode, message: String(cString: errorMessageBuffer))
}

/**
 Verifying proofs using the Groth16 scheme.
 
 - Parameters:
 - proof: The proof data to be verified.
 - inputs: The input data used for verification.
 - verificationKey: The verification key data used for verification.
 - errorBufferSize: The size of the buffer for error messages. Defaults to a predefined value.
 
 - Throws: `RapidsnarkVerifierError` child classes
 if the proof verification fails, with the error message indicating the reason for failure.
 
 - Returns: A boolean value indicating whether the proof is valid (`true`) or not (`false`).
 */
public func groth16Verify(
    proof: Data,
    inputs: Data,
    verificationKey: Data,
    errorBufferSize: Int = defaultErrorBufferSize
) throws -> Bool {
    let proofBuf = proof.nullTerminatedBytes
    let inputsBuf = inputs.nullTerminatedBytes
    let verificationKeyBuf = verificationKey.nullTerminatedBytes
    
    var errorMessageBuffer: [CChar] = Array(repeating: 0, count: errorBufferSize)
    
    // Call the rapidsnark C++ library function to perform the Groth16 Verification
    let statusCode = groth16_verify(
        proofBuf,
        inputsBuf,
        verificationKeyBuf,
        &errorMessageBuffer,
        UInt(errorBufferSize)
    );
    
    if (statusCode == VERIFIER_VALID_PROOF) {
        return true;
    } else if (statusCode == VERIFIER_INVALID_PROOF) {
        throw RapidsnarkVerifierError.invalidProof(message: String(cString: errorMessageBuffer))
    } else if (statusCode == VERIFIER_ERROR) {
        throw RapidsnarkVerifierError.error(message: String(cString: errorMessageBuffer))
    } else {
        throw RapidsnarkUnknownStatusError(message: String(cString: errorMessageBuffer))
    }
}


/**
 Calculates the size of the public signal buffer based on the provided zkey file data.
 
 - Parameters:
 - zkey: The zkey data used to calculate the public signal buffer size.
 - errorBufferSize: The size of the buffer for error messages. Defaults to a predefined value.
 
 - Throws: `RapidsnarkProverError` child classes
 if the calculation of the public signal buffer size fails, with the error message indicating the reason for failure.
 
 - Returns: An integer value representing the calculated size of the public signal buffer.
 */
public func groth16PublicSizeForZkeyBuf(
    zkey: Data,
    errorBufferSize: Int = defaultErrorBufferSize
) throws -> Int {
    let zkeyBuffer = NSData(data: zkey).bytes
    
    var errorMessageBuffer: [CChar] = Array(repeating: 0, count: errorBufferSize)
    
    var size = 0
    
    let statusCode = groth16_public_size_for_zkey_buf(
        zkeyBuffer,
        UInt(zkey.count),
        &size,
        &errorMessageBuffer,
        UInt(errorBufferSize)
    );
    
    if (statusCode == PROVER_OK) {
        return size
    }
    
    throw RapidsnarkProverError.error(message: String(cString: errorMessageBuffer))
}


/**
 Determine the necessary buffer size for storing public signals based on a provided .zkey file
 
 - Parameters:
 - zkeyPath: The path to the .zkey file used to calculate the public signal buffer size.
 - errorBufferSize: The size of the buffer for error messages. Defaults to a predefined value.
 
 - Throws: `RapidsnarkProverError` child classes
 if the calculation of the public signal buffer size fails, with the error message indicating the reason for failure.
 
 - Returns: An integer value representing the calculated size of the public signal buffer.
 */
public func groth16PublicSizeForZkeyFile(
    zkeyPath: String,
    errorBufferSize: Int = defaultErrorBufferSize
) throws -> Int {
    var errorMessageBuffer: [CChar] = Array(repeating: 0, count: errorBufferSize)
    
    var size = 0
    
    let statusCode = groth16_public_size_for_zkey_file(
        zkeyPath,
        &size,
        &errorMessageBuffer,
        UInt(errorBufferSize)
    );
    
    if (statusCode == PROVER_OK) {
        return size
    }
    
    throw RapidsnarkProverError.error(message: String(cString: errorMessageBuffer))
}

// Helper function to convert the status code and message into a specific error type
private func groth16proverStatusCodeErrors(_ statusCode: Int32, message: String) -> RapidsnarkError {
    switch statusCode {
    case PROVER_ERROR:
        return RapidsnarkProverError.error(message: message)
    case PROVER_ERROR_SHORT_BUFFER:
        return RapidsnarkProverError.shortBuffer(message: message)
    case PROVER_INVALID_WITNESS_LENGTH:
        return RapidsnarkProverError.invalidWitnessLength(message: message)
    default:
        return RapidsnarkUnknownStatusError(message: message)
    }
}


public protocol RapidsnarkError : CustomNSError {
    var message: String { get }
}

public extension RapidsnarkError {
    var errorUserInfo: [String : Any] {
        return ["message": message]
    }
}

public enum RapidsnarkProverError : RapidsnarkError {
    case error(message: String)
    case shortBuffer(message: String)
    case invalidWitnessLength(message: String)
    
    public var message: String {
        switch self {
        case .error(let message):
            return message
        case .shortBuffer(let message):
            return message
        case .invalidWitnessLength(let message):
            return message
        }
    }
    
    public var errorCode: Int {
        switch self {
        case .error:
            return NSNumber(value: PROVER_ERROR).intValue
        case .shortBuffer:
            return NSNumber(value: PROVER_ERROR_SHORT_BUFFER).intValue
        case .invalidWitnessLength:
            return NSNumber(value: PROVER_INVALID_WITNESS_LENGTH).intValue
        }
    }
}

public enum RapidsnarkVerifierError : RapidsnarkError {
    case error(message: String)
    case invalidProof(message: String)
    
    public var message: String {
        switch self {
        case .error(let message):
            return message
        case .invalidProof(let message):
            return message
        }
    }
    
    public var errorCode: Int {
        switch self {
        case .error:
            return NSNumber(value: VERIFIER_ERROR).intValue
        case .invalidProof:
            return NSNumber(value: VERIFIER_INVALID_PROOF).intValue
        }
    }
}

public class RapidsnarkUnknownStatusError : RapidsnarkError {
    public let message: String
    
    init(message: String) {
        self.message = message
    }
    
    public var errorCode: Int {
        return -1
    }
}


extension Data {
    var nullTerminatedBytes: UnsafeRawPointer {
        get {
            var data = NSMutableData(data: self)
            
            if data.firstIndex(of: 0) == nil {
                data.append(Data([0]))
            }
            return data.bytes
        }
    }
}

File: android/app/src/main/java/com/example/moproapp/Rapidsnark.kt
package com.example.moproapp

import java.nio.charset.StandardCharsets


const val DEFAULT_PROOF_BUFFER_SIZE = 1024
const val DEFAULT_ERROR_BUFFER_SIZE = 256

// Prover status codes
private const val PROVER_OK = 0x0
private const val PROVER_ERROR = 0x1
private const val PROVER_ERROR_SHORT_BUFFER = 0x2
private const val PROVER_INVALID_WITNESS_LENGTH = 0x3

// Verifier status codes
private const val VERIFIER_VALID_PROOF = 0x0
private const val VERIFIER_INVALID_PROOF = 0x1
private const val VERIFIER_ERROR = 0x2


private val rapidsnarkJNI = RapidsnarkJniBridge()


/**
 * [groth16Prove] is a JNI bridge to the C library rapidsnark.
 * It takes a PATH to zkey and witness as byte array and returns a proof and public signals as byte arrays.
 * This function is used to generate groth16 proofs.
 * @param zkeyPath - path to zkey file
 * @param witness - witness file as byte array
 * @param proofBufferSize - size of the proof buffer, in most case default value should be ok ()
 * @param publicBufferSize - size of the public buffer, can be calculated with `calculate_public_buffer_size`
 * @param errorBufferSize - 256 should be enough
 * @return [ProveResponse] with proof and public signals
 * @throws [RapidsnarkProverError] or [RapidsnarkUnknownStatusError]
 */
@Throws(RapidsnarkProverError::class)
fun groth16Prove(
    zkeyPath: String,
    witness: ByteArray,
    proofBufferSize: Int = DEFAULT_PROOF_BUFFER_SIZE,
    publicBufferSize: Int? = null,
    errorBufferSize: Int = DEFAULT_ERROR_BUFFER_SIZE
): ProveResponse {
    val publicBufferSizeNonNull =
        publicBufferSize ?: groth16PublicBufferSize(zkeyPath, errorBufferSize)

    // Create buffers to get results
    val proofBuffer = ByteArray(proofBufferSize)
    val publicBuffer = ByteArray(publicBufferSizeNonNull)
    val errorBuffer = ByteArray(errorBufferSize)

    val statusCode = rapidsnarkJNI.groth16ProveWithZKeyFilePath(
        zkeyPath,
        witness, witness.size.toLong(),
        proofBuffer, longArrayOf(proofBufferSize.toLong()),
        publicBuffer, longArrayOf(publicBufferSizeNonNull.toLong()),
        errorBuffer, errorBufferSize.toLong()
    )

    if (statusCode == PROVER_OK) {
        // Convert byte arrays to strings
        val proof = String(proofBuffer, StandardCharsets.UTF_8).trim { it <= ' ' }
        val public = String(publicBuffer, StandardCharsets.UTF_8).trim { it <= ' ' }

        return ProveResponse(proof, public)
    }

    val error = String(errorBuffer, StandardCharsets.UTF_8).trim()
    when (statusCode) {
        PROVER_ERROR -> throw RapidsnarkProverError(error)
        PROVER_ERROR_SHORT_BUFFER -> throw RapidsnarkProverError.ShortBuffer(error)
        PROVER_INVALID_WITNESS_LENGTH -> throw RapidsnarkProverError.InvalidWitnessLength(error)
        else -> throw RapidsnarkUnknownStatusError(error)
    }
}

/** [groth16Verify] is a JNI bridge to the C library rapidsnark.
 * Verifies a proof and returns true or false.
 * @param proof - proof as string
 * @param inputs - public signals as string
 * @param proof - proof as string
 * @return - true if proof is valid, false otherwise
 * @throws [RapidsnarkVerifierError] if proof is invalid or error occurred
 */
@Throws(RapidsnarkVerifierError::class)
fun groth16Verify(
    proof: String,
    inputs: String,
    verificationKey: String,
    errorBufferSize: Int = DEFAULT_ERROR_BUFFER_SIZE
): Boolean {
    val errorBuffer = ByteArray(errorBufferSize)

    val result = rapidsnarkJNI.groth16Verify(
        proof,
        inputs,
        verificationKey,
        errorBuffer,
        errorBufferSize.toLong()
    )

    if (result == VERIFIER_VALID_PROOF) {
        return true
    }

    val error = String(errorBuffer, StandardCharsets.UTF_8).trim()
    when (result) {
        VERIFIER_INVALID_PROOF -> throw RapidsnarkVerifierError.InvalidProof(error)
        VERIFIER_ERROR -> throw RapidsnarkVerifierError(error)
        else -> throw RapidsnarkUnknownStatusError(error)
    }
}

/**
 * [groth16PublicBufferSize] is a JNI bridge to the C library rapidsnark.
 * Calculates the size of the public buffer for a given zkey file.
 * In production better to use hardcoded values or cashed values, because the calculation is slow.
 * @param zkeyPath - path to zkey file
 * @param errorBufferSize - 256 should be enough
 * @return - size of the public buffer
 * @throws [RapidsnarkProverError]
 */
@Throws(RapidsnarkProverError::class)
fun groth16PublicBufferSize(
    zkeyPath: String,
    errorBufferSize: Int = DEFAULT_ERROR_BUFFER_SIZE
): Int {
    val errorBuffer = ByteArray(errorBufferSize)

    val publicBufferSize = rapidsnarkJNI.groth16PublicSizeForZkeyFile(
        zkeyPath,
        errorBuffer,
        errorBufferSize.toLong(),
    ).toInt()

    val error = String(errorBuffer, StandardCharsets.UTF_8).trim()

    return publicBufferSize
//    if (error.isEmpty()) {
//        return publicBufferSize
//    } else {
//        throw RapidsnarkProverError(error)
//    }
}

abstract class RapidsnarkError(
    message: String? = null,
    cause: Throwable? = null,
    val code: Int,
) : Exception(message, cause)

open class RapidsnarkProverError(
    message: String? = null,
    cause: Throwable? = null,
    code: Int = PROVER_ERROR,
) : RapidsnarkError(message, cause, code) {
    class ShortBuffer(message: String) :
        RapidsnarkProverError(message, code = PROVER_ERROR_SHORT_BUFFER)

    class InvalidWitnessLength(message: String) : RapidsnarkProverError(
        message, code = PROVER_INVALID_WITNESS_LENGTH
    )
}

open class RapidsnarkVerifierError(
    message: String? = null,
    cause: Throwable? = null,
    code: Int = VERIFIER_ERROR,
) :
    RapidsnarkError(message, cause, code) {
    class InvalidProof(message: String) :
        RapidsnarkVerifierError(message, code = VERIFIER_INVALID_PROOF)
}

class RapidsnarkUnknownStatusError(message: String? = null, cause: Throwable? = null) :
    RapidsnarkError(message, cause, 1)

data class ProveResponse(
    val proof: String,
    val publicSignals: String
)


private class RapidsnarkJniBridge {
    companion object {
        init {
            System.loadLibrary("rapidsnark")
        }
    }

    external fun groth16ProveWithZKeyFilePath(
        zkeyPath: String,
        wtnsBuffer: ByteArray, wtnsSize: Long,
        proofBuffer: ByteArray, proofSize: LongArray,
        publicBuffer: ByteArray, publicSize: LongArray,
        errorMsg: ByteArray, errorMsgMaxSize: Long
    ): Int

    external fun groth16Verify(
        proof: String,
        inputs: String,
        verificationKey: String,
        errorMsg: ByteArray,
        errorMsgMaxSize: Long
    ): Int

    external fun groth16PublicSizeForZkeyFile(
        zkeyPath: String,
        errorMsg: ByteArray,
        errorMsgMaxSize: Long
    ): Long
}